lib/functions.php

```diff
@@ -752,9 +752,20 @@ function coroutine(callable $func) {
  * Upon resolution the Generator return value is used to succeed the promised result. If an
  * error occurs during coroutine resolution the returned promise fails.
  *
- * @param \Generator $generator The generator to resolve as a coroutine
+ * @param \Generator|callable $generator A generator or callable the returns a generator to resolve as a coroutine
+ * @return \Amp\Promise
  */
-function resolve(\Generator $generator) {
+function resolve($generator) {
+    if (is_callable($generator)) {
+        $generator = call_user_func($generator);
+
+        if (!$generator instanceof \Generator) {
+            throw new \LogicException('Callable passed to resolve() did not return an instance of Generator');
```

Comment:
Is that a LogicException or an InvalidArgument? Technically we have not provided a callable returning an instance of Generator, so should be invalid arg?